### DISCUSSION TITLE
Enhancement: possibility of introducing (regio-)randomness

### DISCUSSION THREAD

**Discussion by ricalessandri** [marrink-lab/polyply_1.0#231] [Ideas]
Not really an issue but rather a potential enhancement: the idea is to have the possibility of introducing (regio-)randomness on the polymer sequence. For example, for P3HT, what we model usually is regioregular P3HT. However, "regiorandom" P3HT is also a thing - see https://www.semanticscholar.org/paper/Organic-thin-film-transistors.-Klauk/5301e5839c6d3eb3fc63dc79e28bb7b6fbb313f0/figure/5. Do you think this can be implemented easily? Of course, the two ways of attaching P3HT - i.e., 2  blocks - would be needed. But then the crucial point would be: could then `polyply`, if passed a flag `-random` generate a random copolymer? In the case of P3HT that I have in mind it'd be regiorandom P3HT, however this can be thought about more generally in terms of random copolymers made of different chemical blocks.

**Comment by fgrunewald** [comment]
Thanks for raising the issue! Yes indeed this is very useful and polyply can already do it. You need to invoke the sequence generation program first and generate a .json sequence file that specifies the sequence of your polymer and encodes the randomness. Say you have two P3HT monomers called P3HT and P3HTR for the reverse monomer. P3HT->P3HT is the region-regular order and a mix of P3HT and P3HTR is regio random. Then you'd generate a sequence graph in the following way:

```
polyply gen_seq -from_string A:n_monomers:1:P3HT-0.5,P3HTR-0.5 -seq A -o P3HT.json -name P3HT
```
The 'from_string' generates a single block, which has the label A is just a label. The n_monomers the number of monomers, 1 is the number of branches i.e. it is a linear chain, and then there follows a list of syntax <monomer-name>-probability,<monomer-name>-probability etc. In this case you get a random chain where each monomer has 0.5 probability of being either P3HT or P3HTR. Then using '-seq' you specify the sequence of possible blocks. In this case there is only one block so only.

The generated .json file contains the residue graph of the polymer and can be used with the itp generation tool as follows:
```
polyply gen_itp -seqf P3HT.json -name P3HT -o P3HT.itp -f <input-files>
```
Note that of course you need to specify the links and blocks appropriately.

**Comment by ricalessandri** [comment]
Pretty cool!

To make sure I understood the blocks and links all right, I first tried a PEO-PE random block copolymer. So, I've done:
```
polyply gen_seq -from_string A:10:1:PEO-0.5,PE-0.5 -seq A -o PEOPE.json -name PEOPE
```
followed by:
```
polyply gen_itp -seqf PEOPE.json -name  PEOPE -o PEOPE.itp -lib martini2_polymers
```
However, to obtain the correct `itp` below, I think another link definition is needed (see #98 ), otherwise no `; PEO-PE` bonds would be generated:
```
; polyply-itp

[ moleculetype ]
PEOPE 1

[ atoms ]
 1 C1  1 PE  CH2  1 0.0 45.0
 2 EO  2 PEO COC  2 0.0 45.0
 3 C1  3 PE  CH2  3 0.0 45.0
 4 C1  4 PE  CH2  4 0.0 45.0
 5 EO  5 PEO COC  5 0.0 45.0
 6 C1  6 PE  CH2  6 0.0 45.0
 7 C1  7 PE  CH2  7 0.0 45.0
 8 EO  8 PEO COC  8 0.0 45.0
 9 EO  9 PEO COC  9 0.0 45.0
10 C1 10 PE  CH2 10 0.0 45.0

[ bonds ]
 8  9 1 0.322 7000
 3  4 1 0.46 2000
 6  7 1 0.46 2000

; PE-PEO
 2  3 1 0.45 7000
 5  6 1 0.45 7000
 9 10 1 0.45 7000

; PEO-PE
 1  2 1 0.45 7000
 4  5 1 0.45 7000
 7  8 1 0.45 7000

[ exclusions ]
 3  4
 6  7
```
right?

**Comment by fgrunewald** [comment]
Hi yes that is right. The current M2 links are not complete except for the homopolymers and some selected links I used for testing PEGylated lipids and non-ionic surfactants.

**Comment by fgrunewald** [comment]
@ricalessandri Is this issue taken care of? I think in principle we can do regio randomness and such.

**Comment by ricalessandri** [comment]
Yes, it's taken care of indeed - so feel free to close it. We could perhaps have a reminder somewhere that we can add an example about this to the wiki/tutorials (or maybe no need to put a reminder as it'll be anyway part of the P3HT showcase tutorial).
